File: node.js/백준/Gold/14621. 나만 안되는 연애/나만 안되는 연애.js
const fs = require("fs");
const input = fs
  .readFileSync("/dev/stdin")
  .toString()
  .split("\n");

let idx = 0;

const [n, m] = input[idx++].split(" ").map(Number);;
const genders = input[idx++].split(" ");

const parents = Array.from({ length: n }, (_, index) => index);

const routes = [];
for (idx; idx < input.length; idx++) {
  routes.push(input[idx].split(" ").map((c) => Number(c)));
}
routes.sort((a, b) => a[2] - b[2]);

function union(a, b) {
  const [rootA, rootB] = [find(a), find(b)];
  if (rootA !== rootB) {
    parents[rootB] = rootA;
  }
}

function find(x) {
  let parent = parents[x];
  while (parents[parent] !== parent) {
    parent = parents[parent];
  }
  return parent;
}

let answer = 0,
  cnt = 0;

for (const path of routes) {
  const [u, v, d] = path;

  if (genders[u - 1] !== genders[v - 1] && find(u - 1) !== find(v - 1)) {
    union(u - 1, v - 1);
    answer += d;
    cnt++;
  }
}

console.log(cnt === n - 1 ? answer : -1);

File: Python/백준/Gold/14621. 나만 안되는 연애/나만 안되는 연애.py
import sys
input = sys.stdin.readline


def union(u, v):
    root_u, root_v = find(u), find(v)
    if root_u != root_v:
        parents[root_v] = root_u


def find(x):
    parent = parents[x]
    while parent != parents[parent]:
        parent = find(parent)
    
    return parent


n, m = map(int, input().split())
gender = input().split()

routes = []
for _ in range(m):
    routes.append(list(map(int, input().split())))
routes.sort(key=lambda x: x[2])

parents = [i for i in range(n)]
answer = 0
cnt = 0

for route in routes:
    u, v, d = route
    if gender[u-1] != gender[v-1] and find(u-1) != find(v-1):
        union(u-1, v-1)
        answer += d
        cnt += 1

print(answer if cnt == n - 1 else -1)
